lib/trino-parquet/src/main/java/io/trino/parquet/predicate/PredicateUtils.java

```diff
@@ -253,147 +214,9 @@ private static Map<ColumnDescriptor, Long> getColumnValueCounts(PrunedBlockMetad
-            return Optional.empty();
-        }
-        DictionaryPageHeader dictionaryHeader = pageHeader.getDictionary_page_header();
-        if (dictionaryHeader.getNum_values() > MAX_DICTIONARY_SIZE) {
-            return Optional.empty();
-        }
-        return Optional.of(new PageHeaderWithData(
-                pageHeader,
-                inputStream.readSlice(pageHeader.getCompressed_page_size())));
-    }
-
-    private static DictionaryPage decodeDictionaryPage(ParquetDataSourceId dataSourceId, PageHeaderWithData pageHeaderWithData, ColumnChunkMetadata chunkMetaData, Optional<FileDecryptionContext> decryptionContext)
+    private record PredicateMatchResult(boolean matchesWithoutDictionary, TupleDomainParquetPredicate indexPredicate, Set<ColumnDescriptor> candidateColumnsForDictionaryMatching)
     {
-        PageHeader pageHeader = pageHeaderWithData.pageHeader();
-        DictionaryPageHeader dicHeader = pageHeader.getDictionary_page_header();
-        ParquetEncoding encoding = getParquetEncoding(Encoding.valueOf(dicHeader.getEncoding().name()));
-        int dictionarySize = dicHeader.getNum_values();
-
-        Slice compressedData = pageHeaderWithData.compressedData();
-        Slice maybeDecrypted = compressedData;
-        Optional<ColumnDecryptionContext> columnContext = decryptionContext.flatMap(context -> context.getColumnDecryptionContext(chunkMetaData.getPath()));
-        if (columnContext.isPresent()) {
-            byte[] aad = AesCipherUtils.createModuleAAD(
-                    columnContext.get().fileAad(),
-                    ModuleType.DictionaryPage,
-                    chunkMetaData.getRowGroupOrdinal(),
-                    chunkMetaData.getColumnOrdinal(),
-                    -1);
-            byte[] plain = columnContext.get().dataDecryptor().decrypt(compressedData.getBytes(), aad);
-            maybeDecrypted = Slices.wrappedBuffer(plain);
-        }
-        try {
-            return new DictionaryPage(decompress(dataSourceId, chunkMetaData.getCodec().getParquetCompressionCodec(), maybeDecrypted, pageHeader.getUncompressed_page_size()), dictionarySize, encoding);
-        }
-        catch (IOException e) {
-            throw new ParquetDecodingException("Could not decode the dictionary for " + chunkMetaData.getPath(), e);
-        }
-    }
-
-    private record PageHeaderWithData(PageHeader pageHeader, Slice compressedData)
-    {
-        private PageHeaderWithData(PageHeader pageHeader, Slice compressedData)
-        {
-            this.pageHeader = requireNonNull(pageHeader, "pageHeader is null");
-            this.compressedData = requireNonNull(compressedData, "compressedData is null");
-        }
+        private static final PredicateMatchResult FALSE_WITH_DICT_MATCHING_NOT_NEEDED = new PredicateMatchResult(false, null, null);
+        private static final PredicateMatchResult TRUE_WITH_DICT_MATCHING_NOT_NEEDED = new PredicateMatchResult(true, null, null);
```

Comment:
nit: Can we avoid null here ? Pass `TupleDomain#all` and an empty set ?

Reply:
It's `TupleDomainParquetPredicate` so needs be converted to equivalent of `TupleDomain#all`. Made `indexPredicate` and `candidateColumnsForDictionaryMatching` `Optional` instead which removes wrapping them in Optional later when passing to `RowGroupInfo`